constraints.txt

```diff
@@ -0,0 +1,31 @@
+# Specifies Python package versions for development and testing.
+# To generate:
+# python3 -m venv venv3 && source venv3/bin/activate
+# python3 -m pip install -e .[dev,tests] && pip freeze
```

Comment:
nit: It seems like `dev` extras weren't used here because things like `tox` and its dependencies aren't pinned.

I think it's fine if they're not pinned for now since the motivation behind this issue was pinning our test dependencies, but it'd be good to properly document how this file was created.
```suggestion
# python3 -m pip install -e .[tests] && pip freeze
```

Reply:
Ah, I should have added a comment -- that's necessary to get things working locally for me, so I left it to make things easier when running locally in the future. My intent was that people would run these commands, look at versions of things, and cherry-pick updates as necessary. I intentionally removed the dev-only dependencies since it looked like testing doesn't need those (I think it's installing tox itself and such that makes it necessary for me, and since this is used after installing tox in the tests, that's not necessary). This file isn't the output of those commands directly at all, since some versions needed to be changed, and also after adding the other ones you suggest. I'll add a comment to this effect if you're cool with that?

Reply:
Sure. I personally don't have too strong of an opinion as long as our instructions here are accurate.